axiom/nr_php_packages.c

```diff
@@ -0,0 +1,158 @@
+#include "util_strings.h"
+
+nr_php_package_t* nr_php_package_create(char* name, char* version) {
+  nr_php_package_t* p = NULL;
+
+  if (NULL == name) {
+    return NULL;
+  }
+
+  p = (nr_php_package_t*)nr_malloc(sizeof(nr_php_package_t));
+  if (NULL == p) {
+    return NULL;
+  }
+
+  p->package_name = nr_strdup(name);
+  if (NULL != version) {
+    p->package_version = nr_strdup(version);
+  } else {
+    p->package_version
+        = nr_strdup(" ");  // if null, version is set to an empty
+                           // string with a space according to spec
+  }
+
+  return p;
+}
+
+void nr_php_package_destroy(nr_php_package_t* p) {
+  if (NULL != p) {
+    nr_free(p->package_name);
+    nr_free(p->package_version);
+    nr_free(p);
+  }
+}
+
+void nr_php_packages_add_package(nr_php_packages_t** h, nr_php_package_t* p) {
+  nr_php_package_t* package;
+  if (NULL == h) {
+    return;
+  }
+
+  if (NULL == p || NULL == p->package_name || NULL == p->package_version) {
+    return;
+  }
+
+  if (NULL == *h) {
+    *h = nr_hashmap_create((nr_hashmap_dtor_func_t)nr_php_package_destroy);
+  }
+
+  // If package with the same key already exists, we will check if the value is
+  // different. If it is different, then we will update the value of the package
+  package = (nr_php_package_t*)nr_hashmap_get(*h, p->package_name,
+                                              nr_strlen(p->package_name));
+  if (NULL != package) {
+    if (0 != nr_strcmp(package->package_version, p->package_version)) {
+      nr_free(package->package_version);
+      package->package_version = nr_strdup(p->package_version);
+    }
+    nr_php_package_destroy(p);
+    return;
+  }
+
+  nr_hashmap_set(*h, p->package_name, nr_strlen(p->package_name), p);
+}
+
+size_t nr_hashmap_count_buckets(const nr_php_packages_t* hashmap) {
+  return (size_t)(1 << hashmap->log2_num_buckets);
+}
+
+char* nr_php_package_to_json(nr_php_package_t* package) {
+  char* json;
+  nrbuf_t* buf = NULL;
+
+  if (NULL == package || NULL == package->package_name
+      || NULL == package->package_version) {
+    return NULL;
+  }
+  buf = nr_buffer_create(0, 0);
+  nr_buffer_add(buf, NR_PSTR("[\""));
+  nr_buffer_add(buf, package->package_name, nr_strlen(package->package_name));
+  nr_buffer_add(buf, NR_PSTR("\",\""));
+  nr_buffer_add(buf, package->package_version,
+                nr_strlen(package->package_version));
+  nr_buffer_add(buf, NR_PSTR("\""));
+
+  /* unclear if this empty dictionary should be included or not */
+  nr_buffer_add(buf, NR_PSTR(",{}]"));
+  nr_buffer_add(buf, NR_PSTR("\0"));
+
+  json = nr_strdup(nr_buffer_cptr(buf));
+  nr_buffer_destroy(&buf);
+
+  return json;
+}
+
+bool nr_php_packages_to_json_buffer(nr_php_packages_t* h, nrbuf_t* buf) {
+  size_t num_buckets;
+  size_t i;
+  nr_hashmap_bucket_t* bucket;
+  bool package_added = false;
+  char* package_json;
+
+  if (NULL == h || NULL == buf) {
+    return false;
+  }
+
+  num_buckets = nr_hashmap_count_buckets(h);
```

Comment:
```suggestion
  num_buckets = nr_hashmap_count(h);
```

The consumer of the `nr_hashmap_t` data type does not need to concern itself with the internal structure of the data structure, in this case, the number of buckets in use to implement the hash map.  The correct call is `nr_hashmap_count()` which returns the number of items in the hashmap.